spark3/src/main/java/org/apache/iceberg/spark/source/SparkMicroBatchStream.java

```diff
@@ -0,0 +1,314 @@
+import org.apache.iceberg.util.PropertyUtil;
+import org.apache.iceberg.util.SnapshotUtil;
+import org.apache.iceberg.util.TableScanUtil;
+import org.apache.spark.api.java.JavaSparkContext;
+import org.apache.spark.broadcast.Broadcast;
+import org.apache.spark.sql.connector.read.InputPartition;
+import org.apache.spark.sql.connector.read.PartitionReaderFactory;
+import org.apache.spark.sql.connector.read.streaming.MicroBatchStream;
+import org.apache.spark.sql.connector.read.streaming.Offset;
+import org.apache.spark.sql.util.CaseInsensitiveStringMap;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.apache.iceberg.TableProperties.SPLIT_LOOKBACK;
+import static org.apache.iceberg.TableProperties.SPLIT_LOOKBACK_DEFAULT;
+import static org.apache.iceberg.TableProperties.SPLIT_OPEN_FILE_COST;
+import static org.apache.iceberg.TableProperties.SPLIT_OPEN_FILE_COST_DEFAULT;
+import static org.apache.iceberg.TableProperties.SPLIT_SIZE;
+import static org.apache.iceberg.TableProperties.SPLIT_SIZE_DEFAULT;
+
+public class SparkMicroBatchStream implements MicroBatchStream {
+  private static final Logger LOG = LoggerFactory.getLogger(SparkMicroBatchStream.class);
+
+  private final JavaSparkContext sparkContext;
+  private final Table table;
+  private final boolean caseSensitive;
+  private final Schema expectedSchema;
+  private final Long splitSize;
+  private final Integer splitLookback;
+  private final Long splitOpenFileCost;
+  private final boolean localityPreferred;
+  private final StreamingOffset initialOffset;
+
+  SparkMicroBatchStream(JavaSparkContext sparkContext,
+                        Table table, boolean caseSensitive, Schema expectedSchema,
+                        CaseInsensitiveStringMap options, String checkpointLocation) {
+    this.sparkContext = sparkContext;
+    this.table = table;
+    this.caseSensitive = caseSensitive;
+    this.expectedSchema = expectedSchema;
+    this.localityPreferred = Spark3Util.isLocalityEnabled(table.io(), table.location(), options);
+
+    long tableSplitSize = PropertyUtil.propertyAsLong(table.properties(), SPLIT_SIZE, SPLIT_SIZE_DEFAULT);
+    this.splitSize = Spark3Util.propertyAsLong(options, SparkReadOptions.SPLIT_SIZE, tableSplitSize);
+
+    int tableSplitLookback = PropertyUtil.propertyAsInt(table.properties(), SPLIT_LOOKBACK, SPLIT_LOOKBACK_DEFAULT);
+    this.splitLookback = Spark3Util.propertyAsInt(options, SparkReadOptions.LOOKBACK, tableSplitLookback);
+
+    long tableSplitOpenFileCost = PropertyUtil.propertyAsLong(
+        table.properties(), SPLIT_OPEN_FILE_COST, SPLIT_OPEN_FILE_COST_DEFAULT);
+    this.splitOpenFileCost = Spark3Util.propertyAsLong(options, SPLIT_OPEN_FILE_COST, tableSplitOpenFileCost);
+
+    InitialOffsetStore initialOffsetStore = InitialOffsetStore.getInstance(table, checkpointLocation);
+    this.initialOffset = getOrWriteInitialOffset(initialOffsetStore);
+  }
+
+  @Override
+  public Offset latestOffset() {
```

Comment:
It looks like we expose all available data at once. This may be too much to be processed in one micro batch. I know Spark 3.1 added a limit API. Do we plan to leverage it once we switch to 3.1?

Reply:
Essentially, we need a limit for how much data we should expose in a single micro-batch.

Reply:
yes, this is a limitation with spark's current `MicroBatchStream` interface. it is asking the streaming source only for 2 values: `latest` and `initial` - & is building a MicroBatch out of it. let me explore the limit API a little bit more - but, yes this is very good call out - we must leverage the limit API - if it solves that problem - as soon as we switch to 3.1.

Reply:
I am resolving the comments - the ones which I felt - where we are able to drive reasonable consensus or is tracked in my PR **What next** section - so that - **the ones that need more discussion will pop out**. Pl. definitely reopen or start a new comment - if you feel - the need for more discussion right here.

Reply:
I would prefer to take a different approach here. I agree with Anton that there's a problem creating a batch that is much too large for Spark to process. Fixing that incrementally is a good idea, but I think that the first pass should be to read incrementally from the oldest snapshot rather than reading the entire snapshot. That avoids hitting the size problem in most cases and is probably a better default behavior anyway.

Reply:
@rdblue - can you pl. elaborate on the proposal.
I couldn't understand how your proposal overcomes the hurdle / spark's `MicroBatchStream` implementation limitation - where: Spark is collecting only the `latestOffest` and `initialOffset` from the streaming source implementers. 

Is the proposal that - instead of returning the actual iceberg table's `latestOffset` - return only a manageable amount of data's corresponding offset as the `latestOffset` ?

Reply:
If you don't want to implement rate limiting, then the default should not be to process the entire table to start. Instead, it should be to process only incremental changes in the oldest snapshot. We can add the ability to process the entire table after rate limiting is implemented.

Reply:
That was my initial approach. But then, there is an issue with correctness - in the case of `Trigger.Once`. Seemslike [there is precedence ](https://databricks.com/blog/2017/05/22/running-streaming-jobs-day-10x-cost-savings.html) in using `Trigger.Once` in the context of automating Batch processing - where `readStream.Trigger(Trigger.Once)` is expected to return all available events.

Due to the above reason - we cannot implement incrementally emitting the `latestOffsets`.
& then, evaluating the concern that @aokolnychyi's & you (@rdblue's) raised - that ppl who will try this on their existing tables - which will be most of the usecases - will all be very large tables & will run into the problem of Spark cluster not being able to handle their first batch.

After sleeping on this - I conclude that - implementing rate limiting in the PR is the best way forward.
Pl. let me know your thoughts.

Reply:
From our offline conversation - you **clarified** my question above; I am merging both these threads here - to democratize the info:
> I’m not saying that we should limit `latestOffset`. I’m saying that the starting offset should not set `scanAllFiles` unless the **user explicitly opts in**. That way the table picks up just incremental changes and not the full table data.

So, to put the option in perspective, let me run thru your suggestion with an example.

Lets say that we have an iceberg table at hand - with 4 snapshots to start with. We don't have a streaming reader on this table yet.
```
S1 [dataOpType=Append, AddedFiles=100]
S2 [dataOpType=Append, AddedFiles=100, ExistingFiles=100]
S3 [dataOpType=Append, AddedFiles=100, ExistingFiles=200]
S4 [dataOpType=Append, AddedFiles=100, ExistingFiles=300]
```

lets say that the **Snapshots S1 and S2 got collected - due to retention policy**. So, our iceberg table now has only 2 Snapshots & lookslike:
```
S3 [dataOpType=Append, AddedFiles=100, ExistingFiles=200]
S4 [dataOpType=Append, AddedFiles=100, ExistingFiles=300]
```

Now, lets say that - I am starting a **fresh `STREAM_READ` on this table**. 

### My current implementation will generate the offsets to stream full table:
initialOffset: [SnapshotId: S3, Position: 0, scanAllFiles = **true**] // so returns AddedFiles + ExistingFiles
latestOffset: [SnapshotId: S4, Position: 100, scanAllFiles = false]

**So, in the current implementation - the number of files that the Streaming reader will read here are: 400 - i.e., full table data.**

### The recommendation is to change the default behavior to return added files only:
initialOffset: [SnapshotId: S3, Position: 0, scanAllFiles = **false**] // so returns AddedFiles ONLY
latestOffset: [SnapshotId: S4, Position: 100, scanAllFiles = false]

**So, after this change - the number of files that the Streaming reader will read here are: 200 - i.e., the data that was incrementally appended to the table.**

### My concern
We are balancing the tradeoff between 2 factors here:
1) Overwhelm spark with full table data in the first Micro_Batch (vs)
2) Default behavior - to NOT return full table data.

To me, default behavior to NOT return full table data - seems like **data loss** for streaming scenarios. Did that make sense? I am streaming all data from the table - if the table had all the snapshots - I have returned 400 files - if few snapshots were collected due to retentionPolicy - although the table has all 400 files - we only return 200 files!

**Do you think that the correct way instead would be to, provide an option - to opt out of `scanAllFiles` in the first Micro_batch - if a consumer doesn't care about it?**

Reply:
based on our offline discussion with @rdblue - we landed on changing the default behavior to NOT `scanAllFiles` in first snapshot.
I will introduce an option - if ppl want to start the stream with full scan.
Flag name: `stream-start-mode = full | incremental`

Reply:
**rationale**: streaming an **year old or a multi-year old iceberg table** - could potentially bring a frustrating experience.
So, we will by default support `stream-start-mode` to be `incremental` and provide an option to specify `full`.

Reply:
@rdblue - changed the default scan mode to be `incremental` and added an item to support `fullScan` mode in the **WhatNext** section of this PR - as it might not make much sense to add it at this point - without having rate limiting capability..